src/wallet/scriptpubkeyman.h

```diff
@@ -203,13 +205,20 @@ class ScriptPubKeyMan
 
     virtual const CKeyMetadata* GetMetadata(const CTxDestination& dest) const { return nullptr; }
 
-    virtual std::unique_ptr<SigningProvider> GetSigningProvider(const CScript& script) const { return nullptr; }
+    virtual std::unique_ptr<SigningProvider> GetPublicSigningProvider(const CScript& script) const { return nullptr; }
 
-    /** Whether this ScriptPubKeyMan can provide a SigningProvider (via GetSigningProvider) that, combined with
-      * sigdata, can produce a valid signature.
+    /** Whether this ScriptPubKeyMan can provide a SigningProvider (via GetPublicSigningProvider) that, combined with
+      * sigdata, can produce solving data.
       */
     virtual bool CanProvide(const CScript& script, SignatureData& sigdata) { return false; }
 
+    /** Creates new signatures and adds them to the transaction. Returns whether all inputs were signed */
+    virtual bool SignTransaction(CMutableTransaction& tx, const std::map<COutPoint, Coin>& coins, int sighash, std::map<int, std::string>& input_errors) const { return false; }
+    /** Sign a message with the given script */
+    virtual SigningError SignMessage(const std::string& message, const PKHash& pkhash, std::string& str_sig) const { return SigningError::OK; };
```

Comment:
It would be better to return some error from here (like the neighboring methods do). In case somebody forgets to override this method, it would be unexpected to return OK and leave the output `str_sig` unmodified.

Maybe even use a pure virtual method `...str_sig) const = 0;` to cause a compilation error if somebody forgets to override `SignMessage()`. I checked that:
* it compiles if changed to pure virtual and
* it does not compile if changed to pure virtual and the `LegacyScriptPubKeyMan` child class forgets to override `SignMessage()`.

Reply:
Changed it to return `SigningError::SIGNING_FAILED`. I'm avoiding pure virtual functions in `ScriptPubKeyMan` because there is no requirement for every `ScriptPubKeyMan` to implement every function.